### DISCUSSION TITLE
[Android] Blending animations

### DISCUSSION THREAD

**Discussion by wes-nz** [google/filament#4754] [Q&A]
I have 2 animations inside my GLTF file. I am able to successfully play them. However I am struggling to to blend them together. From reading [an other thread](https://github.com/google/filament/discussions/4228) it should to be possible? 

I am applying 2 animations at a specific time like so

```Kotlin
animator?.applyAnimation(0, 1f)
animator?.updateBoneMatrices()

animator?.applyAnimation(1, 1f)
animator?.updateBoneMatrices()
```

The animation 1 is overriding the animation 0.  

![final](https://user-images.githubusercontent.com/2945380/138204171-d2f105ed-e346-4be8-afbb-baf4347a6515.png)

I've tried and checked my file with GLTF Viewer, everything works well. 

Is there anything I am missing? 
Thanks 🙏

[GLTF file](https://www.dropbox.com/s/r6us0tbdpdr1npt/animated_gltf.zip?dl=0)

**Comment by bejado** [comment]
I don't think we have a way to blend animations with our `Animator` API. Internally, `updateBoneMatrices()` calls `setBones`, which overrides all previous calls.

**Reply by wes-nz** [reply]
It makes sense. With the limitations of 4 morphs and animations not blending, is there any alternative solutions to get a similar result?

**Reply by jnmaloney** [reply]
@wes-nz I need the same functionality but I haven't tackled it yet. I think that the bone matrices will have to be blended and calculated outside of Filament and then passed to the asset. 
I know the library uses gltfio under the hood if that helps.

**Reply by wes-nz** [reply]
Thanks @jnmaloney. You're right I think, it's likely the way to go. I am trying to explore all the easiest options first.

**Comment by prideout** [comment]
If you do `animator.applyAnimation(0, 1f) ; animator.applyAnimation(1, 1f)`, then gltfio will do the expected thing.  It will apply animation 0, then animation 1.  If the two animations affect a disjoint set of nodes, then the second one will not clobber the first one.

I'm not aware of anything in the GLTF spec that describes blending behavior between animations that affect the same set of nodes.  Mathematically, it is unclear how two node transformations should be mixed together.

I tried your model with the BabylonJS viewer, but it doesn't allow multiple concurrent animations.  Next I tried it with [this](https://gltf-viewer.donmccurdy.com/) ThreeJS viewer, and it does not mix the animations.

**Reply by wes-nz** [reply]
Thanks for taking the time to reply, much appreciated.

When using [the linked viewer](https://gltf-viewer.donmccurdy.com), clicking _play all_ appears to play both animations and mix transformations.

https://user-images.githubusercontent.com/2945380/138803945-621aa998-290c-40ca-a8f8-efe21ad3135d.mov

I could be misunderstanding something, my knowledge in that field is limited. Thanks.

**Reply by jnmaloney** [reply]
> Mathematically, it is unclear how two node transformations should be mixed together.

I know this isn't exactly core to Filamant, and sorry if I'm missing something because it's been a while since I did something like this.

I think that when you apply an animation to an asset, then each vertex is transformed by a weighted linear combination of transforms from assigned bones. 

Then what a blend of animations would do is transform each vertex by a weighted linear combination of (the weighted linear combination of transforms from animation A) and (same for animation B).

**Reply by prideout** [reply]
Thanks, I see that now in the ThreeJS viewer.  Do you know if any other tools support this?  I couldn't find anything about it in the spec, but this would be a useful feature, e.g. I can imagine combining a "walk" animation with a "jump" animation.

This seems like a reasonable feature request although some of the details aren't immediately obvious to me, since glTF supports other types of animation, not just skinning.

**Reply by wes-nz** [reply]
I haven't tried with other tools. If you have any specific tools in mind I am happy to give them a go and check. 

On a similar topic, should I create a feature request to be able to edit more than 4 morph targets? - That's how I ended up using animations instead.

**Reply by romainguy** [reply]
We already have an open feature request for this I believe.

**Reply by wes-nz** [reply]
@romainguy I guess you are talking about [this issue](https://github.com/google/filament/issues/1487) or maybe [this one](https://github.com/google/filament/issues/3726)? From my understanding the CPU method (`asset->getSourceAsset()`) isn't accessible in Java. I don't want to derail the initial conversation further. I will create an enhancement request and we can close it if it's irrelevant.

**Reply by simonsland** [reply]
> We already have an open feature request for this I believe.

which open feature? Is that have already developed?

**Reply by August1996** [reply]
> Thanks, I see that now in the ThreeJS viewer. Do you know if any other tools support this? I couldn't find anything about it in the spec, but this would be a useful feature, e.g. I can imagine combining a "walk" animation with a "jump" animation.
> 
> This seems like a reasonable feature request although some of the details aren't immediately obvious to me, since glTF supports other types of animation, not just skinning.


I have the same use case, is there a good solution for mix animations?

**Reply by Berna78** [reply]
I am also looking for a solution to this question...I have to create a smooth transition from different animation.

**Reply by prideout** [reply]
@Berna78 we definitely have a plan to address this issue for morphing (not skinning), by letting users apply an arbitrary weighting factor at run time.

**Reply by Berna78** [reply]
When could this new feature be included on Filament?

**Reply by August1996** [reply]
@prideout 
This is the solution of babylonjs.
https://playground.babylonjs.com/#6I67BL%232

Refers to the idea, i use filament to do animation transition.

https://user-images.githubusercontent.com/15972489/169214997-1cbf815c-de77-4570-9e1e-9d3c078c2f3c.mp4

**Reply by August1996** [reply]
> @prideout This is the solution of babylonjs. https://playground.babylonjs.com/#6I67BL%232
> 
> Refers to the idea, i use filament to do animation transition.
> 
>  Screenrecorder.mp4

When transiting, i will calculate the blending animation between two animations.
```
float3 scale1;
quatf rotation1;
float3 translation1;
gltfio::decomposeMatrix(xform1, &translation1, &rotation1, &scale1);

float3 scale2;
quatf rotation2;
float3 translation2;
decomposeMatrix(xform2, &translation2, &rotation2, &scale2);

float3 targetScale = lerp(scale1, scale2, weight);
quatf targetRotation = slerp(rotation1, rotation2, weight.x);
float3 targetTranslation = lerp(translation1, translation2, weight);
```

**Reply by Berna78** [reply]
@August1996 Could you share your demo code?

**Comment by prideout** [comment]
I am starting to work on this feature and I want to make a correction.  I said earlier that this will apply to morphing but not skinning.  I had this backwards!  The initial implementation will actually apply only to skinning, and it will be limited to a cross-fade between two animations.
